src/lib/png/png.c

```diff
@@ -0,0 +1,1128 @@
+const int DISTANCE_ADDITIONAL[30]
+	= {0, 0, 0, 0, 1, 1, 2, 2, 3, 3, 4, 4, 5, 5, 6, 6, 7, 7, 8,
+		8, 9, 9, 10, 10, 11, 11, 12, 12, 13, 13};
+
+/* represented by codes 257-285 */
+const int LENGTH_BASE[29]
+	= {3, 4, 5, 6, 7, 8, 9, 10, 11, 13, 15, 17, 19, 23, 27, 31, 35, 43, 51, 59,
+		67, 83, 99, 115, 131, 163, 195, 227, 258};
+
+/* addit bits with codes 257-285 (added to base length) */
+const int LENGTH_ADDITIONAL[29]
+	= {0, 0, 0, 0, 0, 0, 0,  0, 1, 1, 1, 1, 2, 2, 2, 2, 3, 3, 3, 3,
+		4, 4, 4, 4, 5, 5, 5, 5, 0};
+/*
+the order in which "code length alphabet code lengths" are stored, out of this
+the huffman tree of the dynamic huffman tree lengths is generated
+*/
+const int CLCL_ORDER[NUMBER_OF_LENGTH_CODES]
+	= {16, 17, 18, 0, 8, 7, 9, 6, 10, 5, 11, 4, 12, 3, 13, 2, 14, 1, 15};
+
+#define READBIT(bitpointer, bitstream) ((bitstream[bitpointer >> 3] >> (bitpointer & 0x7)) & (unsigned char)1)
+
+/* def Huffmantree struct and functions to work with it */
+typedef struct HuffmanTree {
+	int *tree2d;
+	int *tree1d;
+	int *lengths; /* lengths of the codes(tree1d) */
+	int maxbitlen; /* max bits that single code can get */
+	int num_of_codes; /* number of symbols in the alph */
+} HuffmanTree;
+
+void HuffmanTree_init(HuffmanTree* tree) {
+	tree->tree2d = 0;
+	tree->tree1d = 0;
+	tree->lengths = 0;
+}
+
+void HuffmanTree_cleanup(HuffmanTree* tree) {
+	free(tree->tree2d);
+	free(tree->tree1d);
+	free(tree->lengths);
+}
+
+/*Two functions to convert string of bytes to int */
+int mypow(int digit,int degree) {
+	int i, res = 1;
+	for (i = 0; i < degree; i++) {
+		res = res * digit;
+	}
+	return res;
+}
+
+int convert(unsigned char *bytes, int len) {
+	int i, res = 0;
+	for(i = 0; i < len; i++) {
+		res = res + bytes[i] * mypow(16, (len - 1 - i) * 2) ;
+	}
+	return res;
+}
+
+/* fun to decode symbols, return decoded symbol or error(-1) */
+int huffmanDecodeSymbol(unsigned char* in, int* bp,
+				HuffmanTree* tree, int inbitlength) {
+	int pos = 0, tmp;
+	/* endless 'for' until get symbol or get error */
+	for(;;) {
+		if(*bp >= inbitlength) {
+			return -1; /* error: end of input memory reached without endcode */
+		}
+		tmp = tree->tree2d[(pos << 1) + READBIT(*bp, in)];
+		(*bp)++;
+
+		if(tmp < tree->num_of_codes) {
+			return tmp; /*get decoded symbol, return it*/
+		} else {
+			pos = tmp - tree->num_of_codes; /* not yet symbol, continue */
+		}
+
+		if(pos >= tree->num_of_codes) {
+			return -1; /* error: it appeared outside the codetree */
+		}
+	}
+}
+
+/*
+make from tree1d[] tree2d[][]. value = 6666 means not init
+value < number_of_codes is a codes
+value >= number_of_codes is an address to another bit
+The 2 rows are the 2 possible bit values (0 or 1), there are as
+many columns as codes - 1.
+*/
+int make_tree2d(HuffmanTree *tree) {
+	int nodefill_info = 0;
+	int pos = 0; /*pos in the tree */
+	int i, n;
+
+	tree->tree2d = (int*)malloc(tree->num_of_codes * 2 * sizeof(int));
+	if(!tree->tree2d) {
+		return 66; /* malloc fail */
```

Comment:
What is this magic constant? ENOMEM seems to be more suitable.

Reply:
done in two places